### ISSUE TITLE
Class Distance Loss not working

### ISSUE THREAD

**Issue by mullerpeter** [uber/ludwig#194]
Using `class_distance` and `class_distance_temperature` for Category Output Features always leads to the error "No class_distance metadata available for ...". 
While checking the code I noticed that the `CategoryBaseFeature` class tries to access the class distances via the key `distances` from the features metadata. But they are never stored under that key. I think the whole implementation to calculate and store class distances is missing from the toolkit.

**Comment by w4nderlust** [comment]
@mullerpeter thanks for reporting this. After a quick look, it's likely that it's just a mismatch of names, will set it as a bug and resolve it quickly.
Anyway, there is no calculation for class distances, the user is expected to provide the distances between classes in the model definition, and that is used by Ludwig in the calculation of the loss https://github.com/uber/ludwig/blob/master/ludwig/features/category_feature.py#L293 . Anyway, will use the occasion of fixing this bug to probably enhance this part, maybe renaming distances to similarities for instance, which makes a bit more sense, and adding some additional documentation about it.
Will post the progress in this thread.

**Comment by mullerpeter** [comment]
One thing which I think is particularly unclear in the documentation, how to deal with the `<UNK>` class and ordering when specifying the `class_distances`. From the code i see that the user is supposed to order the distances according to the category to integer ID mapping from the metadata JSON and needs to add the distances to the `<UNK>` class as well. (In contrast for the `class_weights` a 0 gets added automatically for the `<UNK>` class, which is also not mentioned in the documentation) Maybe that information should be added as well.

**Comment by w4nderlust** [comment]
These are both good points. I should clarify both in the docs and maybe make it so that both behave in the same way.

**Comment by w4nderlust** [comment]
@mullerpeter I did several changes in the two commits listed above. I renames `class_distance` to `class_similarities`, fixed bugs that made it so it was not working, explained better in the docs how does it work, adopted the same convention (requiring a value for all classes, including `<UNK>`) for both `class_weights` and `class_similarities`. I also added hopefully more clear error messages.
Let me know if this solves your specific issue and if you believe the issue can be closed.
